lib/Twig/Environment.php

```diff
@@ -213,24 +216,37 @@ public function isStrictVariables()
     }
 
     /**
-     * Gets the cache directory or false if cache is disabled.
+     * Gets the current cache implementation.
      *
-     * @return string|false
+     * @return Twig_CacheInterface|string|false A Twig_CacheInterface implementation,
+     *                                          an absolute path to the compiled templates,
+     *                                          or false to disable cache
      */
     public function getCache()
     {
-        return $this->cache;
+        return $this->originalCache;
     }
 
     /**
-     * Sets the cache directory or false if cache is disabled.
+     * Sets the current cache implementation.
      *
-     * @param string|false $cache The absolute path to the compiled templates,
-     *                            or false to disable cache
+     * @param Twig_CacheInterface|string|false $cache A Twig_CacheInterface implementation,
+     *                                                an absolute path to the compiled templates,
+     *                                                or false to disable cache
      */
     public function setCache($cache)
     {
-        $this->cache = $cache ? $cache : false;
+        $this->originalCache = $cache;
```

Comment:
this should be done only after validating it. Otherwise, someone calling ``setCache`` and catching the exception would be able to set an invalid value in this property.